### ISSUE TITLE
Crash while calibrating steering wheel

### ISSUE THREAD

**Issue by marcmy** [Ryochan7/DS4Windows#688]
Happened 3 times while trying to calibrate. The program just closes without any error or warning. Don't know if there's a log file or it's located. Also seems it won't calibrate properly when I try.

**Comment by mika-n** [comment]
What settings you use with "steering wheel" emulation? Which option as output axis? If it is VJoy axis then you need to install VJoy virtual joystick driver (althrough, it should't crash if the driver is missing. Log file should complain about missing vjoy). If you use any other output axis then vjoy driver is not needed.

When you press "Calibrate..." button then does it crash right away or only after you try to calibrate center/left 90degree/right 90degree tilt positions?

When you press "Calibrate..." button then do you see the calibrate instructions screen? Calibration needs to be done while that dialog is open and press "OK" after you have completed the calibration.
- Place "steering wheel" (=DS4 gamepad) in center position, hold it steady and press "X" button. The red lightbar color changes to yellow when center calibration point is accepted.
- Now turn the controller at 90 degree left position, hold it steady and press "X" button. The lightbar turns to light blue when 90 deg left (or right depending on which way you went from center) calibration point is accepted.
- Now turn the controller at 90 degree opposite position (left or right from center) and press "X". Wait until lightbar turns to green.
- When the color is green then all 3 calibration points are set. You can now keep on turning the controller and check calibration points by looking at when the lightbar starts to blink (=it is hold steady at one of the calibration points).
- If you are happy with the calibration then press "OK" to close the calibration instructions dialogbox.

DS4Win app does (and the calibration process) output some log messages to "Log" options tabpage but you cannot see it if thecalibration is done via "Calibrate..." button. However, there is alternative way to initiate steering wheel calibration. You can assign a special action to a gamepad key, so this way you can first open the "Logs" tabpage of DS4Windows and then press a defined "steering wheel calibration special action trigger key". Now you should  be able to see debug and calibration messages while you are doing the calibration.

Note! Please note that this "steering wheel emulation" mode pretty much requires that DS4 controller is attached to a do-it-yourself rig to make sure that calibration and gyro sensor values are consistent. If you try to do it while holding the controller freely in your hand then the result is probably not that great (ie. gyro sensor values are not consistent compared to calibration points).
https://github.com/Ryochan7/DS4Windows/pull/491#issuecomment-439821245

https://github.com/Ryochan7/DS4Windows/wiki/Settings#steering-wheel

**Comment by marcmy** [comment]
apologies for the delay. It happened with Left X-Axis, range 360. It doesn't crash right away when I click calibrate, i see the instructions. When I try to follow them, it doesn't seem to work, as in the lightbar doesn't blink when turning to the correct angle, and pressing X does nothing. After trying this a few times, the entire program just shuts down.

**Comment by mika-n** [comment]
> Edit: I just read your note and I was trying to do it freely in my hand. So that's probably why it wasn' >working, thanks.

Jep. Probably that's it because the calibration checks the "stability" of the controller and if it shakes too much it doesn't accept the calibration point. Even if you would manage to hold the controller steady enough and calibrate all 3 calibration points then later on if gyro values are not consistent with calibrated points then the result (=angle calculation of the "steering wheel") is not very accurate. 

If you wan't to hold the controller freely in hand then use "gyro steering wheel option 1". But that option has very limited "steering wheel turn range" (max 128 degrees).
https://github.com/Ryochan7/DS4Windows/wiki/Settings#steering-wheel

If you build do-it-yourself-cardboard/wooden rig (or whatever technique) to hold the controller as a "real" steering wheel then the option 2 is much more precise and supports wider ranges. Not to mention that steering wheel immersion will be 100x better.

I guess I have to make it more clear in a documentation that the option-2 is probably usable only with a DIY rig.

**Comment by Ryochan7** [comment]
I tested the calibration routine again and it worked fine on my end. The values from a mounted DS4 would be more consistent but the steering wheel mode still works pretty well when holding the controller in your hands.

I have used it to play ATS even though I haven't been able to play it for a while. Driving is easier with the increased range that the steering wheel mode offers. The biggest difference is during the parking segments. I had a hard time with using the regular tilt controls to park but the steering wheel mode improves the controls so much for parking.
